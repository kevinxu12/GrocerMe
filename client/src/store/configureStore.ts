import { createStore, applyMiddleware, compose, CombinedState } from 'redux';
import createSagaMiddleware from 'redux-saga';
// import createEngine from 'redux-storage-engine-localstorage';
import { createReducer } from './reducers';
import thunk from 'redux-thunk';
// import { reviver } from './localstorage';
import { persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import { RootState } from 'types/RootState';

export function configureAppStore(initialState = {}) {
  /* eslint-disable */
  let composeEnhancers =
    (window['__REDUX_DEVTOOLS_EXTENSION_COMPOSE__'] as typeof compose) ||
    compose;
  const reduxSagaMonitorOptions = {};
  const sagaMiddleware = createSagaMiddleware(reduxSagaMonitorOptions);

  // Create the store with two middlewares
  // 1. sagaMiddleware: Makes redux-sagas work
  // 2. routerMiddleware: Syncs the location/URL path to the state
  // const reducer = storage.reducer(createReducer());
  const rootReducer = createReducer();
  // const engine = createEngine('my-save-key', reviver);
  // const middleware = storage.createMiddleware(engine);
  const middlewares = [sagaMiddleware, thunk]; //middleware];

  const persistConfig = {
    key: 'root',
    storage: storage,
  };

  const pReducer = persistReducer<CombinedState<RootState>, any>(
    persistConfig,
    rootReducer,
  );

  const enhancers = [applyMiddleware(...middlewares)];
  const store = createStore(
    pReducer,
    initialState,
    composeEnhancers(...enhancers),
  );

  // const load = storage.createLoader(engine);
  // load(store);

  // const reduxSagaMonitorOptions = {};
  // const sagaMiddleware = createSagaMiddleware(reduxSagaMonitorOptions);
  // const { run: runSaga } = sagaMiddleware;

  // // Create the store with saga middleware
  // const middlewares = [sagaMiddleware];

  // const enhancers = [
  //   createInjectorsEnhancer({
  //     createReducer,
  //     runSaga,
  //   }),
  // ] as StoreEnhancer[];

  // const store = configureStore({
  //   reducer: createReducer(),
  //   middleware: [...getDefaultMiddleware(), ...middlewares],
  //   devTools:
  //     /* istanbul ignore next line */
  //     process.env.NODE_ENV !== 'production' ||
  //     process.env.PUBLIC_URL.length > 0,
  //   enhancers,
  // });

  return store;
}
