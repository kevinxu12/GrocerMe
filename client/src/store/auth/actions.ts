import { ChangeUsernameAction, LogoutAction } from './../../types/actions';
import { CHANGE_USERNAME, LOGOUT } from './constants';
/**
 * Changes the input field of the form
 *
 * @param  {string} username The new text of the input field
 *
 * @return {object} An action object with a type of CHANGE_USERNAME
 */
export const changeUsername = (username: String): ChangeUsernameAction => ({
  type: CHANGE_USERNAME,
  payload: {
    username,
  },
});

export const logout = (): LogoutAction => {
  return {
    type: LOGOUT,
    payload: {},
  };
};

export default changeUsername;
