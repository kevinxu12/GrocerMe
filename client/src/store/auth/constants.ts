/*
 * HomeConstants
 * Each action has a corresponding type, which the reducer knows and picks up on.
 * To avoid weird typos between the reducer and the actions, we save them as
 * constants here. We prefix them with 'yourproject/YourComponent' so we avoid
 * reducers accidentally picking up actions they shouldn't.
 *
 * Follow this format:
 * export const YOUR_ACTION_CONSTANT = 'yourproject/YourContainer/YOUR_ACTION_CONSTANT';
 */

import { ActionType } from './../../types/actions';

export const CHANGE_USERNAME: ActionType = 'client/store/auth/CHANGE_USERNAME';
export const LOGOUT: ActionType = 'client/store/auth/LOGOUT';
