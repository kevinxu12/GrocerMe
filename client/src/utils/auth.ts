/**
 * @file Generate authentication related helpers
 * @author Kevin Xu
 */

import { Role, RoleCode } from 'types/rest';
import { Constants } from './constants';

export const NO_ROLE_PATHS = [
  Constants.HOME_URL,
  Constants.TEST_SOCKET_URL,
  Constants.ABOUT,
  Constants.CONTACT_US,
];
const ALL_ROLE_PATHS = [Constants.SUPPLIER_HOME, Constants.USER_HOME];
const ADMIN_PATHS = [Constants.ADMIN_HOME];
const CONSUMER_PATHS = [];
const SUPPLIER_PATHS = [];
const role_path_object = {
  ADMIN: ADMIN_PATHS,
  CONSUMER: CONSUMER_PATHS,
  SUPPLIER: SUPPLIER_PATHS,
};

/**
 * Check if a given set of roles is authenticated for a path
 *
 * @param {string} path the path we want to check authentication for
 * @param {Role[]} roles the roles a user is permissioned to
 * @returns {boolean} whether the path is authenticated depending on the roles
 */
export const isPathAuthenticated = (path: string, roles: Role[]): boolean => {
  const role_codes = roles
    .filter((role: Role) => role.status)
    .map((role: Role) => role.code);
  // Admin, Supplier
  let all_paths = role_codes.map(code => role_path_object[code]);
  if (roles.length > 0) {
    all_paths = all_paths.concat(ALL_ROLE_PATHS);
  }
  const all_available_paths: string[] = NO_ROLE_PATHS.concat(
    [].concat(...all_paths),
  );
  return all_available_paths.includes(path);
};

/**
 * Check if a roleCode is contained within a list of roles
 *
 * @param {RoleCode} roleCode RoleCode that may or may not be included
 * @param {Role[]} roles roles to check within
 * @returns {boolean} true if included, false otherwise
 */
export const isRoleCodeIncluded = (
  roleCode: RoleCode,
  roles: Role[] | null,
): boolean => {
  if (!roles) {
    return false;
  }
  const matching_role = roles.filter((role: Role) => role.code === roleCode);
  return matching_role.length > 0;
};
