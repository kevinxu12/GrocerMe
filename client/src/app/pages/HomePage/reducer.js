/*
 * HomeReducer
 *
 * The reducer takes care of our data. Using actions, we can
 * update our application state. To add a new action,
 * add it to the switch statement in the reducer function
 *
 */

import produce from 'immer';
import { Constants } from 'utils/constants';
import { CHANGE_USERNAME } from './constants';

// The initial state of the App
// const storagedata = JSON.parse(localStorage.getItem("data"));
export const initialState = {
  username: Constants.DEFAULT_STRING,
};

/* eslint-disable default-case, no-param-reassign */
const homeReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case CHANGE_USERNAME:
        // Delete prefixed '@' from the github username
        draft.username = action.username;
        break;
    }
  });

export default homeReducer;
