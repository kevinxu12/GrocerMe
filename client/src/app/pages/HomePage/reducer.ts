/*
 * HomeReducer
 *
 * The reducer takes care of our data. Using actions, we can
 * update our application state. To add a new action,
 * add it to the switch statement in the reducer function
 *
 */

import produce from 'immer';
import { HomeAction } from 'types/actions';
import { HomeState } from 'types/RootState';
import { Constants } from 'utils/constants';
import { CHANGE_USERNAME } from './constants';

// The initial state of the App
export const initialState: HomeState = {
  username: Constants.DEFAULT_STRING,
};

/* eslint-disable default-case, no-param-reassign */
const homeReducer = (
  state: HomeState = initialState,
  action: HomeAction,
): HomeState => {
  return produce(state, draft => {
    switch (action.type) {
      case CHANGE_USERNAME:
        draft.username = action.payload.username;
        break;
    }
  });
};

export default homeReducer;
