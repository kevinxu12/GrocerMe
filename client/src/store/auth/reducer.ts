/*
 * HomeReducer
 *
 * The reducer takes care of our data. Using actions, we can
 * update our application state. To add a new action,
 * add it to the switch statement in the reducer function
 *
 */

import produce from 'immer';
import { AuthAction, ChangeUsernameAction } from './../../types/actions';
import { AuthState } from './../../types/RootState';
import { Constants } from './../../utils/constants';
import { CHANGE_USERNAME, LOGOUT } from './constants';

// The initial state of the App
export const initialState: AuthState = {
  username: Constants.DEFAULT_STRING,
};

/* eslint-disable default-case, no-param-reassign */
const authReducer = (
  state: AuthState = initialState,
  action: AuthAction,
): AuthState => {
  return produce(state, draft => {
    switch (action.type) {
      case CHANGE_USERNAME:
        if (!action.payload) {
          break;
        }
        draft.username = (action as ChangeUsernameAction).payload.username;
        break;
      case LOGOUT:
        return initialState;
    }
  });
};

export default authReducer;
