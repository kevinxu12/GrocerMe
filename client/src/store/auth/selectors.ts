import { createSelector } from 'reselect';
import { AuthState, RootState } from './../../types/RootState';
import { initialState } from './reducer';

const selectAuth = (state: RootState): AuthState => state.auth || initialState;

const makeSelectUsername = () =>
  createSelector(
    selectAuth,
    (authState: AuthState): String => authState.username,
  );

export { selectAuth, makeSelectUsername };
