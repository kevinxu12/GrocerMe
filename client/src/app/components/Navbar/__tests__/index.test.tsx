import * as React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { NavBar } from '..';
import { initialState } from 'store/auth/reducer';

jest.mock('react-i18next', () => ({
  useTranslation: () => {
    return {
      t: str => str,
      i18n: {
        changeLanguage: () => new Promise(() => {}),
      },
    };
  },
}));

describe('<Navbar  />', () => {
  const state = { auth: initialState };
  const mockStore = configureStore();
  let store;

  it('should match snapshot', () => {
    store = mockStore(state);
    const loadingIndicator = render(
      <Provider store={store}>
        <NavBar />
      </Provider>,
    );
    expect(loadingIndicator.container.firstChild).toMatchSnapshot();
  });
});
