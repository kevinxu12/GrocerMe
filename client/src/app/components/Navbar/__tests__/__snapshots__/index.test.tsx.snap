// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navbar  /> should match snapshot 1`] = `
.c1 {
  width: 30%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  font-size: 1.2rem;
  font-weight: 400;
  font-size: 32px;
  -webkit-letter-spacing: -1.28px;
  -moz-letter-spacing: -1.28px;
  -ms-letter-spacing: -1.28px;
  letter-spacing: -1.28px;
}

.c3 {
  width: 70%;
  height: 10%;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  margin-left: 40%;
  margin-bottom: 1px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: -1rem;
}

.c5 {
  font-family: Inter;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 18px;
  -webkit-letter-spacing: -0.36px;
  -moz-letter-spacing: -0.36px;
  -ms-letter-spacing: -0.36px;
  letter-spacing: -0.36px;
  font-weight: 500;
  margin-left: 40px;
}

.c5:hover {
  opacity: 0.8;
}

.c5:active {
  opacity: 0.4;
}

.c0 {
  height: 4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: 'rgba(250, 250, 250, 1)' position:fixed;
  top: 0;
  margin-left: 1%;
  width: 100%;
  z-index: 2;
}

.c0 .sc-jSFjdj {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

<header
  class="c0"
>
  <div
    class="c1"
  >
    <h1
      class="c2"
    >
      Grocer + Me
    </h1>
  </div>
  <div
    class="c3"
  >
    <nav
      class="c4"
    >
      <a
        class="c5"
      >
        About us
      </a>
      <a
        class="c5"
      >
        Buy
      </a>
      <a
        class="c5"
      >
        Sell
      </a>
      <a
        class="c5"
      >
        Careers
      </a>
      <a
        class="c5"
      >
        Request early access
      </a>
    </nav>
  </div>
</header>
`;
